### ISSUE TITLE
ReadTheDocs will not build

### ISSUE THREAD

**Issue by roedoejet** [ReadAlongs/Studio#28]
ReadTheDocs cannot install our package which is causing it to fail because autodocumentation tools depend on it. The root cause is because PocketSphinx [fails](https://readthedocs.org/projects/readalong-studio/builds/10738620/) to pip install.

- We could use conda, but we have a number of dependencies that are also not on conda, and we're not currently packaging for conda, so there's some significant work to be done here.
- We could try and fix the pip installation of readalongs, or create a minimal installation without all the audio stuff that crashes it.
- We could spin off our own instance of rtd or generate html locally and push somewhere...but...ugh
- We could abandon autodocumentation...also ugh

Basically there are no great options as far as I'm concerned, but we should try and figure out a solution one way or another.

**Comment by dhdaines** [comment]
Ouf! Help is on the way... I am like 50% complete on a refactored PocketSphinx that won't cause us so many problems...

**Comment by roedoejet** [comment]
Oh that's fantastic! That would solve this entirely 🤘

**Comment by joanise** [comment]
Yes, that's awesome, thanks @dhdaines!

**Comment by dhdaines** [comment]
Actually a question about this - I will need to upload it to PyPI for it to be pip installable right?

I may need some help with Windows building, I have Windows 10 on my laptop but it has some issues...

**Comment by joanise** [comment]
Where to upload it will be a question for Aidan, I've never done such packages. I know how to make something installable locally using `pip install -e .`, but not how to get something onto a proper pip repository.

As for the Windows build, I'll try my best to help, but I don't even have a C or C++ compiler installed yet, so it might take some work. I'm willing to take a stab at it, though, so let me know where things are and what needs to be done.

**Comment by dhdaines** [comment]
Ok thanks!  I am using CMake for building and I have removed all the dependencies and merged PocketSphinx and SphinxBase.  SWIG won't be necessary just to build it, I think.

The idea is basically to have a PocketSphinx that is designed to be embedded on the Web, and abandons any pretense at being LVCSR (because if you want that you should use Kaldi - even the other PocketSphinx developers have switched).  So it does FSG, word spotting, and force alignment only, and I've also removed the fixed-point code since everything has an FPU now (and JavaScript doesn't even really have integer math...) I also want to simplify the API a bit but that might be a larger task.

**Comment by joanise** [comment]
OK, let me know where the repo is when it's ready for me to try compiling. Sounds like you've simplified it enough that it will be easier than it used to be, that's a good thing!

**Comment by roedoejet** [comment]
Yes, so it looks like [Dmitry Prazdnichnov](https://pypi.org/user/bambucha/) is the maintainer of the `pocketsphinx` PyPi package. If you want to publish under the name `pocketsphinx` you'll have to contact him about it. Unfortunately I don't believe PyPi supports [author namespaces](https://github.com/pypa/packaging-problems/issues/114) so you can't just publish `@dhdaines/pocketsphinx` or something like that, which is what the NPM/JavaScript world would have you do.

I definitely recommend using twine for managing your PyPi packages. Once you create your sdist and bdist_wheel files, and [register](https://pypi.org/account/register/) with PyPi, you just `twine upload dist/*` and it manages the rest which is great, and easier than it used to be. I think you will need to pip install twine first.

This all sounds great! Let me know if I can help at all.

**Comment by dhdaines** [comment]
It's no problem, I'll probably change the name, I'm already a bit annoyed that he usurped the pocketsphinx name because there was already a Python module with that name inside the PocketSphinx code - also - I have been annoyed for like 15 years that "Sphinx" is something totally different in the Python world :)

**Comment by roedoejet** [comment]
> It's no problem, I'll probably change the name, I'm already a bit annoyed that he usurped the pocketsphinx name because there was already a Python module with that name inside the PocketSphinx code 

Oh, that **is** annoying!

> - also - I have been annoyed for like 15 years that "Sphinx" is something totally different in the Python world :)

Haha yeah, documentation generation in your pocket? that's all I hear 😂

**Comment by littell** [comment]
Great, I look forward to this.  What functions will be accessible in the API?  I was playing around with pocketsphinx.js but couldn't figure out what was exposed except for the basic functions shown in the tutorial code.

Name-wise, maybe the stripped down version of PocketSphinx is even smaller than a pocket.  MicroSphinx?  PocketSphinxNano?

**Comment by dhdaines** [comment]
I think I'm going to call it SoundSwallower, which is the name of a chapter (as well as a "device" which probably doesn't actually do anything unless you are slowly poisoning yourself with hallucinogenic plants, but that's anohter story) in my current favorite book: https://dflewisreviews.wordpress.com/2016/06/03/datura-or-a-figment-seen-by-everyone/#comment-7561

PocketSphinx.js actually has its own API code implemented in C++ (which I think blows up the size of the JavaScript, though I'm not certain of that) - to call the underlying PocketSphinx API takes some Emscripten magic knowledge

The PocketSphinx decoder API is fairly solid I think, so I will keep it. The audio code is gone, it was a mess of difficult to satisfy platform requirements and dependencies.  The "search switching" API is awkward and kind of badly thought out, as is the configuration (command-line parsing), it would be nice to do away with them...

**Comment by joanise** [comment]
Question: along all the stripped dependencies, are you keeping the ability to convert various file formats using ffmpeg?

**Comment by dhdaines** [comment]
That functionality was actually always external to PocketSphinx itself, so there should be no difference...

**Comment by roedoejet** [comment]
Thanks to @dhdaines this is fixed as of f9b8ce0

Documentation is now building just fine https://readalong-studio.readthedocs.io/en/latest/index.html

**Comment by dhdaines** [comment]
Great news! I'll see if I can fix the Travis build too now :)

**Comment by roedoejet** [comment]
Awesome! I believe the reason why the Travis build was failing was because of a couple tests that were failing. I believe @joanise was going to look into those tests, but I'm not sure... Eric, did you find anything with those?

**Comment by joanise** [comment]
@roedoejet looking for which tests those were, but I don't remember. Can you point me to the right issue?

**Comment by joanise** [comment]
Hum, I can't click through to the Travis CI dashboard anymore, to see what the problem might be. Do I need to create my own account just to see our progress there? I thought it used to work without logging in.

**Comment by dhdaines** [comment]
There were some issues with paths in the Travis script, which I fixed. The only test failure left is in the force alignment test, I will send you the log so you can see it - somehow the word IDs that come out of alignment don't match the ones in the XML file.

Not sure how you can log into Travis CI unfortunately :(

**Comment by joanise** [comment]
The only test in `Studio` that fails on my machine is this one below, but no, I have not had a chance to look into it, sorry.

```
======================================================================
FAIL: testAlignText (test_force_align.TestForceAlignment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\joanise\RAS\Studio\test\test_force_align.py", line 48, in testAlignText
    self.assertEqual(xw.attrib['id'], w['id'])
AssertionError: 't0b0d0p0s0w0' != 's0w0'
- t0b0d0p0s0w0
+ s0w0
```

**Comment by joanise** [comment]
@dhdaines  Thanks. So yes, we're looking at the same test.
I don't particularly want to log into Travis CI, I'd just like the view the dashboard for Studio. It used to work by just clicking on the link in `README.md`.

**Comment by joanise** [comment]
The build and coverage links at the top of the readme don't work for me right now, do they still work for you?

**Comment by dhdaines** [comment]
I think I just fixed them!

Le mer. 22 avr. 2020, à 11 h 05, Eric Joanis <notifications@github.com> a
écrit :

**Comment by joanise** [comment]
Yes, the links work now, thank you!
